cpp/arcticdb/version/version_core.cpp

```diff
@@ -304,92 +324,137 @@ void check_update_data_is_sorted(InputTensorFrame& frame, const index::IndexSegm
         "When calling update, the existing data must be sorted.");
 }
 
-VersionedItem update_impl(
+struct UpdateRanges {
+    IndexRange front;
+    IndexRange back;
+    IndexRange original;
+};
+
+folly::Future<AtomKey> async_update_impl(
     const std::shared_ptr<Store>& store,
     const UpdateInfo& update_info,
     const UpdateQuery& query,
     const std::shared_ptr<InputTensorFrame>& frame,
-    const WriteOptions&& options,
+    WriteOptions&& options,
     bool dynamic_schema,
     bool empty_types) {
-    util::check(update_info.previous_index_key_.has_value(), "Cannot update as there is no previous index key to update into");
-    const StreamId stream_id = frame->desc.id();
-    ARCTICDB_DEBUG(log::version(), "Update versioned dataframe for stream_id: {} , version_id = {}", stream_id, update_info.previous_index_key_->version_id());
-    auto index_segment_reader = index::get_index_reader(*(update_info.previous_index_key_), store);
-    util::check_rte(!index_segment_reader.is_pickled(), "Cannot update pickled data");
-    auto index_desc = check_index_match(frame->index, index_segment_reader.tsd().index());
-    util::check(
-        index_desc.type() == IndexDescriptor::Type::TIMESTAMP || index_desc.type() == IndexDescriptor::Type::EMPTY,
-        "Update not supported for non-timeseries indexes"
-    );
-    check_update_data_is_sorted(*frame, index_segment_reader);
-    bool bucketize_dynamic = index_segment_reader.bucketize_dynamic();
-    (void)check_and_mark_slices(index_segment_reader, dynamic_schema, false, std::nullopt, bucketize_dynamic);
-    fix_descriptor_mismatch_or_throw(UPDATE, dynamic_schema, index_segment_reader, *frame, empty_types);
-
-    std::vector<FilterQuery<index::IndexSegmentReader>> queries =
-        build_update_query_filters<index::IndexSegmentReader>(query.row_filter, frame->index, frame->index_range, dynamic_schema, index_segment_reader.bucketize_dynamic());
-    auto combined = combine_filter_functions(queries);
-    auto affected_keys = filter_index(index_segment_reader, std::move(combined));
-    std::vector<SliceAndKey> unaffected_keys;
-    std::set_difference(std::begin(index_segment_reader),
-                        std::end(index_segment_reader),
-                        std::begin(affected_keys),
-                        std::end(affected_keys),
-                        std::back_inserter(unaffected_keys));
-
-    util::check(affected_keys.size() + unaffected_keys.size() == index_segment_reader.size(), "Unaffected vs affected keys split was inconsistent {} + {} != {}",
+    return index::async_get_index_reader(*(update_info.previous_index_key_), store).thenValue([=](index::IndexSegmentReader&& index_segment_reader) {
```

Comment:
Why are we capturing everything by copy?

Reply:
The whole implementation is inside `async_get_index_reader` and we need to propagate the input params. I used it as a shorthand. In theory, options can be moved but it's consisted only of PODs so it won't do anything.